### ISSUE TITLE
New Relic error: experiment variant conditions not found

### ISSUE THREAD

**Issue by amurphy-cl** [CarnegieLearningWeb/UpGrade#1385]
**Version where bug was found:**
5.1.4

**Describe the bug**
Experiment id: `f3276c73-4c5d-45b9-9462-33a77e9b3818`

New Relic is showing recent errors that variants are not found. Error appears to be in `/mark` call. 

<img width="1456" alt="image" src="https://github.com/CarnegieLearningWeb/UpGrade/assets/65311369/12e863db-987a-4349-bdfa-74cc371b8f7f">

<img width="1493" alt="image" src="https://github.com/CarnegieLearningWeb/UpGrade/assets/65311369/04b47891-346f-4704-beb2-c1e5466b6f91">

**Comment by danoswaltCL** [comment]
Ran across this in load-testing so I was able to actually chase this down:

branch: release/v5.0

When running an experiment at a competing decision point, sometimes the _unassigned_ experiment is wrongly pulled up instead of the experiment they were assigned. The `condition not found` error is because the marked condition payload does not exist in the payload of the incorrect experiment.

Explanation:
The client is allowed to mark the assignment without an experiment id. That will go into this block of code in `ExperimentAssignmentService.markExperimentPoint() line 194

```
      } else {
        const random = seedrandom(userId)();
        experiments = [experiments[Math.floor(random * experiments.length)]];
        experimentId = experiments[0]?.id;
      }
```

That's where something goes wrong. When this error occurs, it is because the experiment that this selects (`experiments[0]`) is the wrong experiment.

When dropping down to 229 we get this block of code:

```
     const experiment = experiments[0]; <----- remember, wrong experiment

      const { conditions } = experiment;
      const payloadCondition = conditions.flatMap((condition) => condition.conditionPayloads);

     <------ the wrong experiment's payloads are searched for the condition, but the competing experiment doesn't have that payload

      const matchedCondition = conditions.filter((dbCondition) => dbCondition.conditionCode === condition);
      const matchedPayloadCondition = payloadCondition.filter((con) => con.payloadValue === condition);

      // console log the variables from the code block above
      logger.info({
        message: `markExperimentPoint: Experiment: ${experimentId}, User: ${userId}, Site: ${site}, Target: ${target}, Condition: ${condition}`,
        details: {
          matchedCondition,
          matchedPayloadCondition,
          conditions,
          payloadCondition,
        },
      });

      if (matchedCondition.length === 0 && matchedPayloadCondition.length === 0 && condition !== null) {
        const error = new Error(`Condition not found: ${condition}`);
        (error as any).type = SERVER_ERROR.CONDITION_NOT_FOUND; 
        logger.error(error);
        throw error;
      }
```

To reproduce:

1. In an empty database, import these segments:
[Segments (1).zip](https://github.com/CarnegieLearningWeb/UpGrade/files/15000579/Segments.1.zip)

2. Then import this experiment:
[MARCS BKT + LLM Rewrites Course 3.json]
(https://github.com/CarnegieLearningWeb/UpGrade/files/15000553/MARCS.BKT.%2B.LLM.Rewrites.Course.3.json)

3. Set to enrolling
4. /init /assign /mark find a user assigned `worksheet_grapher_a1_mod_initial_plus_point` with `cluster_params` or `cluster_params_llm`

5. This should succeed
6. Then import this experiment:

[MARCS Omnibus (3).json](https://github.com/CarnegieLearningWeb/UpGrade/files/15000554/MARCS.Omnibus.3.json)

7. /init /assign /mark that user (or another one with the same conditions above), and this time it should fail and say `no condition found`

In the error logs you should be able to debug and see that `experiments[0]` is not the assigned experiment.
@ppratikcr7  @bcb37

**Comment by danoswaltCL** [comment]
Some follow-up, the conditions for this to occur may have something to do with if they are excluded from one experiment and not the others in a competing decision point, it looks like the `/mark` call is not properly filtering out excluded experiments when marking an assignment without a known experiment id. When using the strategy to find the assignment via the user's userid in the random seed, there is not a guarantee that the list it is selecting "randomly" from will be the same experiment list, in the same order. they do not seem to be sourced from the same query or go through the same filters.

I think by luck this may not have as big an impact as it could, but my worry is that users may be able to mark an experiment that they are not actually supposed to be in. I was not able to verify this one way or another by EOD, it seems like the ones that are actually erroring may be the happy path (nothing happens because it can't mark), but in cases where it succeeds silently to mark a user after "finding" a match like "control" or "variant" in the payload of the wrong experiment's same decision point, I'm not sure what that impact would be.

This is a prod issue and should need as quick a hotfix as we can no matter what the impact has already been.

@VivekFitkariwala @ppratikcr7 @bcb37 @amurphy-cl

**Comment by VivekFitkariwala** [comment]
Which version of client lib are you using? In the newer version of the client lib ie 5, we are sending the experiment ID while marking so that should work fine. I guess it is coming from the earlier versions where we are not sending the experiment ID.

I checked the code and the logic for creating the experiment pool is not the same in the mark and assign call. This could be the reason for this behaviour.

**Comment by danoswaltCL** [comment]
@amurphy-cl We talked about this in morning meeting, Vivek and I came to same conclusions about this bug:

Firstly, neither the Java nor Typescript client libs do not send `experimentId` in mark even in the latest v5. That is likely the immediate prod hotfix to avoid the code where this bug can happen.

It can occur for some users when marking without an experimentId, if and only if they are excluded from one or more experiment that share a decision point
1. The mark call needs to "guess" at which experiment they were in, which requires recreating the strategy used to "randomly" pick an experiment at that shared decision point during assignment, which requires userId and a list of experiments to be exactly the same.
2. The mark call is getting a list of experiments, but not filtering for exclusion, which messes up the integrity of the "guessing" strategy, because it is not the same experiment list.
3. This should only actually happen when a user is at a competing decision point where the are excluded from at least one of the experiments.

There are 3 scenarios that then can happen:
1. the "guessing" strategy ends up picking the right experiment anyway, no harm no foul
2. it picks the wrong experiment, where the "payload" string for that decision point does not match the mark'd value, and it throws an error and does not mark. This is what we see with `no condition found: cluster_params`.
3. it picks the wrong experiment, where the "payload" string for that decision point does actually happen to match (i.e. control condition in both experiments), so the mark succeeds (!) but they are still excluded.

Impact on prod experiments with shared decision points:
So 2) and 3) are what we're worried about, but it may luckily be limited:
For 2), the user will see the right condition, but we will not mark it.
For 3), the user will see the right condition, but mark the wrong experiment, BUT they will still be excluded from it. I have seen this in my local tests, I successfully marked control for a user who was assigned control for a "bkt" experiment, they marked control for the omnibus, but they were effectively excluded from it. Still, a user would encounter the "correct" control condition, so I'm not sure what that impact is totally?

**Comment by danoswaltCL** [comment]
Cancel the "hotfix", the Java 5 lib does in fact send the `experimentId`, I was wrong on that. So does the TS lib. The Java 3 lib does not, which is why we see this error sometimes in prod, and my Python locust tests do not use a lib, and were not originally coded to send an `experimentId` in mark.

So this will be fixed as soon as Mathia 8.18 goes into prod, because that will update our lib from 3 to 5, unless we feel it needs to go sooner.

Follow-up should still be to either make experimentId mandatory, or fix the experiment filter's include/exclude logic in the /mark assignment service call if we still have use cases we need to support where mark may not have that experimentId (there might be, but I can't think of one really....)
